### DISCUSSION TITLE
Generic Inheritance with dynamic behavior

### DISCUSSION THREAD

**Discussion by EricOuellet2** [dotnet/csharplang#6155] [Language Ideas]
My proposition is based on :
https://stackoverflow.com/questions/72319792/does-c-sharp-support-generic-inheritance-and-how-why/72320952?noredirect=1#comment127766253_72320952

I think my  proposition has a potential that could easily be under estimated.

It is a dynamic "Impostor". To keep it as easy to understand I will define it with what properties the class should have:

- A class that could inherits from anything, either generic ones (by opposition of DynamicObject)
- A class that would be recognise as being of its base class at design time (like any regular inheritance)
- A class where all call to any methods and properties would be redirected to a single "Invoke" methods
- The class has one real member implementation by default, a method: "Invoke". The Invoke Method receive all calls to any members.  The default behavior of "Invoke" redirect calls to there appropriate base members or throw a DynamicOperationException
- We can override that "Invoke" method which would enable us to do logging, implements notifications, track changes, ...
- We Should be able to define a class that inherits from "Impostor" in order to add Interface support like INotifyPropertyChanged

- The main Tricks here, is that as decribed, it would require multiple inheritance. That's where Microsoft should be creative. EIther by using multiple inheritance for adding dynamic behavior or by using special attribute, or anything else. That's where the fun begin!

Results:
- A class when used would terribly slowdown performance
- Open a world of almost no limits !!!!! The paradise class !!!!

Please let me know what you think about it? 
Please consider my proposition like it would come from Anders Hejlsberg 😁 and answer the same :-)!!

**Update 2022-05-27**

```
// This is not real code. My intention was only a way to illustrate what I had in mind to, I hope, makes thing easier to show my mind.
// T Could be a specific class like any base class.
public **dynamicClass** MyDynamicClass<T> : T, INotifyPropertyChanged
{
	public event PropertyChangedEventHandler? PropertyChanged;
        public event PropertyChangingEventHandler? PropertyChanging;
	
	public override object PropertyGet(PropertyInfo pi, T obj, object?[]? index )
	{
		// Can log activity
		object result = pi.GetValue(obj);
		return result;
	}

	public override void PropertySet(PropertyInfo pi, T obj, object value, object?[]? index)
	{
		// Can log activity
		// Can track changes
		PropertyChanging?.(obj, ...);
		object result = pi.SetValue(obj, value);
		PropertyChanged?.(obj, ...);
	}

	public override object MethodInvoke(MethodInfo mi, T obj, object[] params)
	{
		// Can log activity
		return mi.Invoke(obj, params);
	}
	
	public override IEnumerable<Attribute> GetCustomAttributes(T obj)
	{
		return obj.GetType().GetCustomAttributes(); // Can add more Attributes
	}
	
	// There is only one way to create a dynamic class. Only one constructor with 1 argument. The argument is an existing instance of any class.
	public MyDynamicClass(T obj)
	{
		base(obj);
	}
}
```

- A dynamicClass is a class with very specific behavior that would makes all inherited class members like they are virtual.
- A dynamicClass class can replace any existing class and can easily add (generic) behaviors.
- A dynamicClass class has default method behavior that can be overwritten. It is like a very special class that could inherits from a base class and act has a derived class, **PLUS**, at the same time, add some specific methods that would make it dynamic (like a late binding class or a class with an IDispatch interface). 
- Added methods are fixed. It is like a dynamicClass would have herited of 2 classes at the same time. One of the 2 is a regular inheritance while the second one has fixed members that brings the "dynamic" behaviors to it.
- A dynamicClass class would funnel all T methods and T properties to a single member that can be overwriten, either those private and not marked as virtual will be funnelled to a single method which act like an IDispatch (LateBinding).
- A dynamicClass class cannot has T being a class marked "sealed".
- Using a dynamicClass would result in similar behavior as using T. But it would add behaviors that would not be possible otherwise.
- A dynamicClass is available at design time and can be used instead of any other class (replacing expected target class) for binding in a WPF or WinUI application.
- The prototype of a dynamicClass, although able to add dynamic behavior is defined by its inherited class and it's own class member. Only methods and property behavior could be enhanced.
 
Note: The implementation in the Language could also be made differently than the one shown. For a second example (other suggestion) :
```
[DynamicClass]
public class MyDynamicClass<T> : T, INotifyPropertyChanged
...

```

Thank you very much!
Eric

**Comment by huoyaoyuan** [comment]
Extending types in such way sounds like extensions discussed in #5496

**Reply by EricOuellet2** [reply]
As far as what I understand from [#5496](https://github.com/dotnet/csharplang/discussions/5496), it add new members but it does not enable us to ovewrite existing member behavior. What I look for is like converting an early binding type to a late binding type. I mean I want to add an equivalent of "IDispatch" to any type where it still continue to provide it's Type definition and entry in a vtable but all calls are redirected to a single point method "Invoke" where we can add anything for each member call like logging, notification, etc...

**Reply by EricOuellet2** [reply]
I will try to create a graphic to better explain my idea (suggestion)
